File: clio/src/main/java/co/casterlabs/packetproto/io/NetCalc.java
package co.casterlabs.packetproto.io;

import lombok.Getter;
import lombok.NonNull;

@Getter
public class NetCalc {
    private int requiredBytes = 0;

    public NetCalc allocateByte() {
        this.requiredBytes++;
        return this;
    }

    public NetCalc allocateBytes(int bytes) {
        this.requiredBytes += bytes;
        return this;
    }

    public NetCalc allocateString(@NonNull String str) {
        byte[] bytes = UTF8StringIOUtil.stringToBytes(str);

        // STR_SIZE + STR_BYTES... + NULL
        this.allocateInt();
        this.requiredBytes += bytes.length;
        this.allocateNull();

        return this;
    }

    public NetCalc allocateNull() {
        this.requiredBytes++;
        return this;
    }

    public NetCalc allocateLong() {
        this.requiredBytes += Long.BYTES;
        return this;
    }

    public NetCalc allocateInt() {
        this.requiredBytes += Integer.BYTES;
        return this;
    }

    public NetCalc allocateShort() {
        this.requiredBytes += Short.BYTES;
        return this;
    }

    public NetCalc allocateBoolean() {
        this.requiredBytes++;
        return this;
    }

    public NetCalc allocateDouble() {
        this.requiredBytes += Double.BYTES;
        return this;
    }

    public NetCalc allocateFloat() {
        this.requiredBytes += Float.BYTES;
        return this;
    }

}

File: clio/src/main/javascript/NetCalc.mjs
class NetCalc {

    constructor() {
        this.requiredBytes = 0;
    }

    allocateByte() {
        this.requiredBytes++;
        return this;
    }

    allocateBytes(len) {
        this.requiredBytes += len;
        return this;
    }

    allocateString(str) {
        const utf8Encoder = new TextEncoder();

        const bytes = utf8Encoder.encode(str);

        this.allocateInt32();
        this.requiredBytes += bytes.length;
        this.allocateNull();

        return this;
    }

    allocateNull() {
        this.requiredBytes++;
        return this;
    }

    allocateBigInt64() {
        this.requiredBytes += 8;
        return this;
    }

    allocateInt32() {
        this.requiredBytes += 4;
        return this;
    }

    allocateInt16() {
        this.requiredBytes += 2;
        return this;
    }

    allocateBoolean() {
        this.requiredBytes++;
        return this;
    }

    allocateFloat64() {
        this.requiredBytes += 8;
        return this;
    }

    allocateFloat32() {
        this.requiredBytes += 4;
        return this;
    }

}

export default NetCalc;